taskwarrior_syncall/tw_asana_utils.py

```diff
@@ -0,0 +1,160 @@
+            as_due_at = parse_datetime(tw_due)
+        else:
+            as_due_at = tw_due
+        as_due_on = as_due_at.date()
+
+    if not isinstance(tw_modified, datetime.datetime):
+        as_modified_at = parse_datetime(tw_modified)
+    else:
+        as_modified_at = tw_modified
+
+    as_name = tw_description
+
+    # Build Asana task
+    return AsanaTask(
+        completed=as_completed,
+        completed_at=as_completed_at,
+        created_at=as_created_at,
+        due_at=as_due_at,
+        due_on=as_due_on,
+        modified_at=as_modified_at,
+        name=as_name,
+    )
+
+
+def convert_asana_to_tw(asana_task: AsanaTask) -> TwItem:
+    # Extract Asana fields
+    as_completed = asana_task["completed"]
+    as_completed_at = asana_task["completed_at"]
+    as_created_at = asana_task["created_at"]
+    as_due_at = asana_task["due_at"]
+    as_due_on = asana_task["due_on"]
+    as_modified_at = asana_task["modified_at"]
+    as_name = asana_task["name"]
+
+    # Declare Taskwarrior fields
+    tw_completed = None
+    tw_due = tw_item = None
+    tw_end = None
+    tw_entry = None
+    tw_modified = None
+    tw_status = "pending"
+
+    # Convert Asana fields to Taskwarrior fields
+    if isinstance(as_created_at, datetime.datetime):
+        tw_entry = as_created_at
+    else:
+        tw_entry = parse_datetime(as_created_at)
+
+    if as_completed:
+        if isinstance(as_completed_at, datetime.datetime):
+            tw_end = as_completed_at
+        else:
+            tw_end = parse_datetime(as_completed_at)
+        tw_status = "completed"
+
+    if as_modified_at is not None:
+        if isinstance(as_modified_at, datetime.datetime):
+            tw_modified = as_modified_at
+        else:
+            tw_modified = parse_datetime(as_modified_at)
+
+    # Asana may give us either 'due_at' (contains time and zone) or 'due_on'
+    # (contains neither).
+    if as_due_at is not None:
+        if isinstance(as_due_at, datetime.datetime):
+            tw_due = as_due_at
+        else:
+            tw_due = parse_datetime(as_due_at)
+    elif as_due_on is not None:
+        if isinstance(as_due_on, datetime.date):
+            tw_due = datetime.datetime.combine(
+                as_due_on, datetime.time(0, 0, 0), dateutil.tz.tzlocal()
+            )
+        else:
+            tw_due = parse_datetime(as_due_on)
+
+    tw_description = as_name
+
+    tw_task = {
+        "description": tw_description,
+        "end": tw_end,
+        "entry": tw_entry,
+        "modified": tw_modified,
+        "status": tw_status,
+    }
+
+    if tw_due is not None:
+        tw_task["due"] = tw_due
+
+    if tw_modified is not None:
+        tw_task["modified"] = tw_modified
+
+    return tw_task
+
+
+def list_asana_workspaces(client: asana.Client) -> None:
+    """List the Asana workspaces known to the provided client."""
+    items = {}
+
+    workspaces = client.workspaces.find_all()
```

Comment:
See my comment in the PR main page but iterating this seems to make an API call per item which doesnt' really scale (will take a long while to complete if the user has 100s of items in Asana.

* Is there a client method to fetch these in bulk - maybe setting the `page_size` to a much higher value might help? https://github.com/Asana/python-asana/#options=

Reply:
Thanks for calling this out.

> but iterating this seems to make an API call per item

This is actually intentional.

https://github.com/bergercookie/taskwarrior-syncall/pull/52/files#diff-e32b17c14d925106b95e4bc6e38235a0fcee593ce06f14948d89eaf08e7e3177R35

The Asana API gives us high-level task info with `.find_all()`, but in order to get details like `due_at` we need to make follow-up API call for each task.

I will increase `page_size` because that's a good idea, but I'm not sure how much faster it will make things.

Some thoughts on other things we can do to improve....

We can make use of API filters `--completed-since` and `--modified-since`. For example:

 1. The first time we sync a combination, we can use `--completed-since=now`, which will only return incomplete tasks. 
     * Maybe the user wants to pull down completed tasks too, in which case we can give users a flag to opt-in to that
 2. Subsequent syncs can use the `--modified-since=<last-sync-time>` API
 3. Every sync will trigger saving `<last-sync-time>` to disk somewhere.

The one major issue with this idea is that the current framework will delete items if they aren't returned by `.get_items()`. Probably a simple way to deal with this would be to add a `--ignore-asana-deletes` option to the framework not to process deletes from the Asana side. This isn't ideal, but users could mitigate this by doing a periodic full sync.

Reply:
> We can make use of API filters --completed-since and --modified-since.

Yeah, agree, both to this and then to defaulting to the modified time 👍🏾 

> Maybe the user wants to pull down completed tasks too, in which case we can give users a flag to opt-in to that

Agree, let's do this!

>  The one major issue with this idea is that the current framework will delete items if they aren't returned by .get_items(). Probably a simple way to deal with this would be to add a --ignore-asana-delete

Sounds good..
Indeed, currently `item_synchronizer` it will flag it as deleted if it's not returned by the `get_all_items`, which isn't always what one may want. This seems like a functionality that other sync scripts may want, so when adding the flag for `--ignore-asana-deletes`, I'd create a more generic function that accepts the name of the side that it wants to create the flag for, similar to, `opt_combination(name_A: str, name_B: str)` but with a single parameter instead of two.

Reply:
I haven't reworked how tasks are fetched, and if it's alright with you will left that out of this PR, but I did raise the page size from 50 to 100 (the max size allowed by the API).